crates/nu-cli/src/repl.rs

```diff
@@ -44,16 +45,51 @@ const PRE_EXECUTE_MARKER: &str = "\x1b]133;C\x1b\\";
 // const CMD_FINISHED_MARKER: &str = "\x1b]133;D;{}\x1b\\";
 const RESET_APPLICATION_MODE: &str = "\x1b[?1l";
 
+///
+/// This shared stack exists to make sure that we never explicitly copy the stack.
+///
+trait SharedStack {
+    fn convert_env_values(&mut self, engine_state: &mut EngineState) -> Option<ShellError>;
+    fn add_env_var(&mut self, var: String, value: Value);
+    fn get_env_var(&mut self, envine_state: &EngineState, var: &str) -> Option<Value>;
+    fn child_stack(&self) -> Stack;
+    fn get_unique_mut(&mut self) -> &mut Stack;
+}
+
+impl SharedStack for Arc<Stack> {
+    fn get_unique_mut(&mut self) -> &mut Stack {
+        let stack = Arc::get_mut(self).expect("More than one reference to the stack found!");
+        return stack;
+    }
```

Comment:
This feels really important for the REPL: we want to make sure we're not accidentally triggering copies when we end up mutating the top-level stack.